handler/incoming_instance_handler.go

```diff
@@ -115,26 +117,52 @@ func (hdlr *InstanceEventHandler) insertDimensions(instance *model.Instance, dim
 
 	cache := make(map[string]string)
 
+	var wg sync.WaitGroup
+	problem := make(chan error, len(dimensions))
+	count := 0
+
 	for _, dimension := range dimensions {
+		wg.Add(1)
+		count++
+		go func(d *model.Dimension) {
+			defer wg.Done()
+
+			d, err := hdlr.Store.InsertDimension(context.Background(), cache, instance, d)
+			if err != nil {
+				problem <- errors.Wrap(err, "error while attempting to insert dimension")
+				return
+			}
 
-		dimension, err := hdlr.Store.InsertDimension(context.Background(), cache, instance, dimension)
-		if err != nil {
-			return errors.Wrap(err, "error while attempting to insert dimension")
-		}
+			if err = hdlr.DatasetAPICli.PutDimensionNodeID(instance.InstanceID, d); err != nil {
+				problem <- errors.Wrap(err, "DatasetAPICli.PutDimensionNodeID returned an error")
+				return
+			}
 
-		if err = hdlr.DatasetAPICli.PutDimensionNodeID(instance.InstanceID, dimension); err != nil {
-			return errors.Wrap(err, "DatasetAPICli.PutDimensionNodeID returned an error")
-		}
+			// todo: remove this temp hack once the time codelist / input data has been fixed.
+			if d.DimensionID != "time" {
+				if err = hdlr.Store.CreateCodeRelationship(context.Background(), instance, d.Links.CodeList.ID, d.Option); err != nil {
+					problem <- errors.Wrap(err, "error attempting to create relationship to code")
+					return
+				}
+			}
+		}(dimension)
 
-		// todo: remove this temp hack once the time codelist / input data has been fixed.
-		if dimension.DimensionID != "time" {
-			if err = hdlr.Store.CreateCodeRelationship(context.Background(), instance, dimension.Links.CodeList.ID, dimension.Option); err != nil {
-				return errors.Wrap(err, "error attempting to create relationship to code")
+		if count >= hdlr.BatchSize {
+			wg.Wait()
+			if len(problem) > 0 {
+				return <-problem
```

Comment:
If my understanding is correct if there are errors will we only ever be returning the first one? If so is that right? Do we want all of the failures? Or do we not care - one or more failures mean the batch has failed? If that is the case I would consider adding `log.Error` before each time we send an err to the `problem` channel to give us visibility of every thing going on.

Reply:
I was going on the theory that the first error is the instigator of any failures, so from a returning perspective, the first one we care about, and 100 errors might be difficult to handle. Logging before passing to the channel makes sense tho ill have a look 👍